Cargo.toml

```diff
@@ -16,6 +16,7 @@ members = [
   "bitmex-stream",
   "xtra-libp2p",
   "xtra-libp2p-ping",
+  "xtra-libp2p-offer",
```

Comment:
Given the amount of dependencies, I'm not sure whether extracting this protocol into a separate crate was a good move (contrary to xtr-libp2p-ping for example, which is standalone and readily reusable in different projects).

I guess we might try it out and see how we go, there's no harm in folding it later (or extract others) if we decide too.

It seems like the `itchysats/model` dependency is the troublesome one (it pulls in all sorts of stuff, even 2 versions of `maia`!), perhaps if we decided to keep it standalone, we could create `MakerOffers` locally and a conversion functions in the daemon?

Reply:
It requires a bit of boilerplate to extract a crate, but I don't see the downside as long as they are standalone. This new crate is independent of all the code in `daemon`, so it seems unnecessary to burden it with all that code when iterating on it.

> It seems like the itchysats/model dependency is the troublesome one (it pulls in all sorts of stuff, even 2 versions of maia!), perhaps if we decided to keep it standalone, we could create MakerOffers locally and a conversion functions in the daemon?

Thank you for caring about the dependency tree! I think that could be a great idea. I'm happy to look into this before merging.

Reply:
I would love to see some benchmarks to prove this - I suspect we might actually be slower due to linking time (and this one will hit us more often than the compilation from scratch - due to incremental compilation it's not really an issue).

I'm not too happy about the flat structure of the general and really specific crates being mixed on the top level of our workspace :/ I'm not sure how to mitigate this though.

overall, I don't think the theoretical upsides are worth the time invested!

Reply:
> Another note on the motivation to extract this into a crate:
> 
> AFAIK, if, for example, we extract two libp2p actors into crates, and the graph looks like this
> 
> ```
>         +- offer -+
>        /            \
> model-+              +-> daemon
>        \            /
>         +- order -+
> ```
> 
> not only are we speeding up the development cycle when working on `offer` and `order` in isolation, but we are also tapping into the compiler's ability to parallelise the compilation of `offer` and `order` when working on `daemon`!
> 
> Given that, I would recommend to extract `rollover` into a separate crate.

It's actually more like this:

```
        +- offer -+
       /            \
model-+  -------     +-> daemon
       \            /
        +- order -+

Reply:
Yes, removing the dependency between the actors and `model` would allow us to parallelise the build of `model`, `offer` and `order`.

That doesn't take away from the fact that building `model`, then `offer` and `order` in parallel, then `daemon`, should be faster than `model` -> `daemon`.

Reply:
> Yes, removing the dependency between the actors and `model` would allow us to parallelise the build of `model`, `offer` and `order`.
> 
> That doesn't take away from the fact that building `model`, then `offer` and `order` in parallel, then `daemon`, should be faster than `model` -> `daemon`.

again, I'd love to see some results. given we're having incremental compilation, I don't think this is worth the time investment.
I feel that it's a premature optimisation, and I'm not too keen on it due to all the boilerplate needed, lack of consistency in pattern usage, lack of reusability of a crate (for me this actually constitutes a crate), and polluting top-level workspace directory.

Reply:
I think it's enough of a gain that one can iterate on the standalone crate without having to build the huge `daemon` crate.
My other argument is just icing on top (and a generally good approach to reduce workspace compile times overall).

> I feel that it's a premature optimisation

If we don't exercise good habits, we're gonna pay the price down the line. We don't want to have to prioritise reducing compilation times for one of our future iterations in a few months. We have been appending things to `daemon` by default because it is very easy to do. I think extracting this crate is just a small step in the right direction.

>  all the boilerplate needed

A `Cargo.toml` file? Even the exercise of extracting this into a crate allows us to think about our dependencies and start questioning them. For example, you came up with considering if we even need to depend on `model`. Had I defaulted to adding all this to `daemon`, we may have never had that thought.

> lack of consistency in pattern usage (unless all of them are migrated)

I am not gonna be convinced to keep adding things to `daemon` just because it is "what we do". We have to start somewhere.

> lack of reusability of a crate

We could reuse these actors differently if we ever decide to rewrite `itchysats`.

> polluting top-level workspace directory

You could argue the same thing about the top-level of the `daemon` crate.

Reply:
Sorry, but I think what you perceive as "good habits" is quite arbitrary. I am not for this change precisely because I think it's not a change in a good direction in terms of structure. 

Can we please not conflate adding new features with arguing about workplace structure? It feels that it mandates a separate discussions / proposals / meeting and *benchmarks*.

One last argument for this: The recent moves of **half** of the files from daemon caused so much wasted time for me personally due to the new lack of consistency introduced that I can guarantee that I wasted more that I'd get back with a clean compilation.

Reply:
> Sorry, but I think what you perceive as "good habits" is quite arbitrary. I am not for this change precisely because I think it's not a change in a good direction in terms of structure.
>
> Can we please not conflate adding new features with arguing about workplace structure? It feels that it mandates a separate discussions / proposals / meeting and benchmarks.

We can take this discussion elsewhere. I have simply been replying to your comment on the PR. I thought it was important to address it because, well, I like to address all feedback. And because I'm pretty sure this is a good change.

> One last argument for this: The recent moves of half of the files from daemon caused so much wasted time for me personally due to the new lack of consistency introduced that I can guarantee that I wasted more that I'd get back with a clean compilation.

I would like to hear more about what exactly was the problem there, because I can't relate.

Reply:
> I would like to hear more about what exactly was the problem there, because I can't relate.

one actor system etc being *inside* the daemon, and one outside. inconsistency that trips me every time!

Reply:
It's true that we haven't done the same thing for the `taker` yet, so it's not ideal. Would it help if I just did it so that we can have consistency again?

The split is not arbitrary, those are actors and code that only run on the `maker` binary.

Reply:
> It's true that we haven't done the same thing for the `taker` yet, so it's not ideal. Would it help if I just did it so that we can have consistency again?
> 
> The split is not arbitrary, those are actors and code that only run on the `maker` binary.

I never said that split was arbitrary, I believe it has merits. I only stated that because of a half-done change we are lacking consistency and it is objectively harder to find stuff, because the same actors/structs for different binaries are in completely different crates. I find this situation unfortunate as I often spend much more time trying to find stuff ever since it went out-of-sync.

Please keep in mind that I am not *against* refactors such like that, I would like to just minimise the state of flux if possible, because otherwise we run the risk of a lingering half-done refactor that makes the codebase less maintainable and will be a nightmare for newcomers to understand
I believe  separating larger refactors from feature work (this is everything I'm trying to argue for in this ticket) is one way to do it. Then we can at least create a ticket,  discuss and implement them, potentially parallelising the work :-)
it would be also easier to benchmark by applying/reverting such a commit.

Reply:
I can accept that the change you are referring to wasn't ideal. With it I was hoping to move the codebase in a better direction (subjective or not this is what I believed at the time and still believe). I prefer to work on refactors in patches before or after introducing features because otherwise it is very hard to find time to do them. 

What I take away from this discussion is that the particular refactor we are talking about was controversial and should have been done at a different time, ideally in full (including the taker side).

I honestly cannot relate to the struggles you are voicing. But I can understand that others might have a different experience to mine. And it's objectively inconsistent when it could easily be consistent.

Now, the change proposed here has little to do with all that. It's not a refactor and it doesn't move any files around that could confuse anyone. I have explained why I think it's a good choice, but at this stage I no longer care.

If anyone is keen to take this over they are welcome to cram it into the `daemon` crate. I'm unassigning myself because I won't be able to push this forward for the next iteration.

Reply:
Let's not throw away the baby with the bathwater here: 
-  I have never said that this other example was in bad direction, I just said it left things in inconsistent state for ages (potentially forever, as there are no TODOs for that, nor issues on github). That example was only pointed out briefly to showcase the downsides of inconsistency, nothing else - my apologies if it derailed the conversation and made it longer than it could have been. I agree with that direction and would be happy to share the load in rolling it out completely.
- doing refactors before or after features is great, I believe this is the right way to go in order not to be overwhelmed by technical debt.

What I had a problem with was doing structural changes amongst the same PR if they could be done in separate steps, especially when extracting it required more work. (and now backtracking to daemon would require even more, so it's not that obvious which path to take!). Perhaps we just have different definitions of what a refactor is, or a threshold for confusion is different (e.g. for me, 1 of the 4 itchysats-specific protocols being outside the daemon is confusing, really). 

Anyway, the work done by you here is awesome - both the protocol implementation along with the comprehensive test suite. I am saddened that we dragged this discussion past the point of caring - I take it that I failed conveying my reasons, and should have backtracked earlier.

Reply:
I think we have actually reached an agreement on the extraction of part of `daemon` into `maker` being subpar (not the idea necessarily, but certainly the execution). Moving forward I will be more careful to not introduce major inconsistencies because it's absolutely true that they can be a bit confusing, particularly to new devs.

I shouldn't have said that I no longer care. It was an emotional response that I should have kept to myself. Sorry about that. The problem is that it's gonna be much harder to extract other protocols and I don't want to keep pushing for it if it's gonna result in annoying inconsistencies and we don't have consensus.

For that reason I would actually prefer it if we put this work back into `daemon` at this stage.